### DISCUSSION TITLE
Can I deploy nuxt 3 in cpanel ??

### DISCUSSION THREAD

**Discussion by Ali-0111** [nuxt/nuxt#25910] [Questions]
Hi, I needed to deploy my nuxt 3 built file in cpanel(not static one). I decided to use webpack instead vite as i can have the index.html and then I can deploy this in cpanel to have a fully dynamic front end.

here is what i did, but could not built:

nuxt.config.ts:
```
import { defineNuxtConfig } from 'nuxt/config';
import HtmlWebpackPlugin from 'html-webpack-plugin';
import HtmlWebpackInlineSourcePlugin from 'html-webpack-inline-source-plugin';

export default defineNuxtConfig({
  // devtools: { enabled: true },
  // debug: true,
  css: ['~/assets/scss/tailwind.scss'],
  builder: 'webpack',
  webpack: {
    plugins: [
      new HtmlWebpackPlugin({
        filename: 'output.html', // output file name that will be created
        template: 'output-template.html', // template file to use for insertion
        inlineSource: '.(js|css)$', // embed all javascript and css inline
        inject: 'body' // inject script at the end of document body
      })
      new HtmlWebpackInlineSourcePlugin()
    ]
  },
  hooks: {
    // may be needed for successful bundling
    'webpack:config'(configs) {
      configs[0].module.rules.push(
        {
          test: /\.vue$/,
          loader: 'vue-loader'
        },
        {
          test: /\.html$/,
          loader: 'html-loader'
        },
        {
          test: /\.tsx?$/,
          use: 'ts-loader',
          exclude: /node_modules/
        },
        {
          test: /\.css$/,
          use: ['style-loader', 'css-loader']
        },
        {
          test: /\.(png|jpg|gif|webp|svg)$/,
          loader: 'url-loader'
        },
        {
          test: /\.s[ac]ss$/i,
          use: [
            // Creates `style` nodes from JS strings
            'style-loader',
            // Translates CSS into CommonJS
            'css-loader',
            // Compiles Sass to CSS
            'sass-loader'
          ]
        }
      );
    }
  },
  ```
  
  package.json:
  
  ```
    "devDependencies": {
    "@babel/eslint-parser": "^7.23.3",
    "@nuxt/image": "^1.0.0-rc.3",
    "@nuxtjs/color-mode": "^3.3.2",
    "@nuxtjs/eslint-module": "^4.1.0",
    "@nuxtjs/stylelint-module": "^5.1.0",
    "@nuxtjs/tailwindcss": "^6.10.4",
    "eslint": "^8.56.0",
    "eslint-config-prettier": "^9.1.0",
    "eslint-plugin-nuxt": "^4.0.0",
    "eslint-plugin-prettier": "^5.1.3",
    "eslint-plugin-vue": "^9.20.1",
    "html-loader": "^5.0.0",
    "html-webpack-inline-source-plugin": "^1.0.0-beta.2",
    "html-webpack-plugin": "^5.6.0",
    "json-joy": "9.8.0",
    "nuxt": "^3.9.1",
    "nuxt-icon": "^0.6.8",
    "nuxt-swiper": "^1.2.2",
    "postcss-custom-properties": "^13.3.5",
    "prettier": "^3.2.2",
    "sass": "^1.71.1",
    "sass-loader": "^14.1.1",
    "style-loader": "^3.3.4",
    "stylelint": "13.x",
    "stylelint-config-standard": "21.x",
    "stylelint-csstree-validator": "1.x",
    "stylelint-scss": "3.x",
    "ts-loader": "^9.5.1",
    "url-loader": "^4.1.1",
    "vue": "^3.4.10",
    "vue-eslint-parser": "^9.4.0",
    "vue-loader": "^17.4.2",
    "vue-router": "^4.2.5"
  },
  "dependencies": {
    "@nuxt/webpack-builder": "^3.10.2",
    "axios": "^1.6.7",
    "vue-dompurify-html": "^5.0.1"
  }
  ```
  
  Now, I am not sure will i get the result? and what could be the solution.
  
  Your efforts, support and time is highly appreciated.

**Comment by urbgimtam** [comment]
Sorry for being so late, but I've deployed it successfully some time ago. 
Posting as this may help others. (I was bitten by the "experimental.writeEarlyHints" on production myself, but didn't understood until some time later).

Please note: 
1. You need your cpanel to have NodeJS apps installed (not all installations of cPanel have them installed.)
2. You'll need ssh access to the server.
3. Make sure you're not using "experimental.writeEarlyHints" on nuxt.config.ts (this will not work on Passenger websites)

Setup:
1. Go to "Setup Node.js App"
3. Select your node version, put Application mode as "development" and save.
4. Select your application root path (this is where the nuxt root files should be), as well as your application url.
5. Create an application startup file in .cjs format (can be something like "entry.cjs") and put it in the nuxt root. This will be the loader for the Nuxt server. The content of this file should be "import('./.output/server/index.mjs')".
6. Start the app in cPanel.
7. After the app is started, click on "Run NPM install".
8. Open your terminal by ssh, and cd to the root of your nuxt app (cPanel embeeded terminal wont work for larger projects)
9. On the terminal, 'npm run build'.
10. On the cPanel, change application mode to "production". Save and restart app.

Should be working, if all goes well.

**Reply by DeVoresyah** [reply]
Is it possible to do the build process without being in the panel?

**Reply by urbgimtam** [reply]
I've been trying to do that exact same thing, but haven't achieved it yet.

I have a git repository in cPanel, and use a deployment file (.cpanel.yml) to config the deployment, which is really straightforward to use (I've used Plesk in the past, and it works in a similar way)

On that deployment config file, you can define where files should be copied, then run `npm i`, `npm run build` and create the file `tmp/restart.txt' to automatically restart.

However, I can't seem to switch the node environment between development and production on that script (was hoping one could use `export NODE_ENV=development`, should have been enough).

Instead, I need to manually go to 'NodeJS app" panel, change mode from "production" to "development", save the change.
I can then run 'npm install' via ssh.
To build, I need to go back to the "NodeJS app" panel, change mode again from "development" to "production", save the changes.
I can then run "npm run build" via ssh. 

This simple thing is the only point preventing me from automatic update and build on commit...

**Reply by sapna-baniya** [reply]
I am getting 503 service unavailable

**Comment by mohssineAboutaj** [comment]
is there any solution? i got **503** too

**Reply by Yehan20** [reply]
Did you check your node version . i got the same error once i changed it to higher version it  seemed worked .
